database/rrdlabels.c

```diff
@@ -3,201 +3,1138 @@
-{
-    return (label_list_lookup_key(head, key, key_hash) != NULL);
+static int simple_pattern_match_name_and_value_callback(const char *name, void *value, void *data) {
+    struct simple_pattern_match_name_value *t = (struct simple_pattern_match_name_value *)data;
+    RRDLABEL *lb = (RRDLABEL *)value;
+
+    // we return -1 to stop the walkthrough on first match
+    if(simple_pattern_matches(t->pattern, name)) return -1;
+
+    size_t len = RRDLABELS_MAX_NAME_LENGTH + RRDLABELS_MAX_VALUE_LENGTH + 2; // +1 for =, +1 for \0
+    char tmp[len], *dst = &tmp[0];
+    const char *v = lb->value;
+
+    // copy the name
+    while(*name) *dst++ = *name++;
+
+    // add the equal
+    *dst++ = t->equal;
+
+    // add the value
+    while(*v) *dst++ = *v++;
+
+    // terminate it
+    *dst = '\0';
+
+    if(simple_pattern_matches(t->pattern, tmp)) return -1;
+
+    return 0;
 }
 
-int label_list_contains(struct label *head, struct label *check)
-{
-    return label_list_contains_key(head, check->key, check->key_hash);
+bool rrdlabels_match_simple_pattern_parsed(DICTIONARY *labels, SIMPLE_PATTERN *pattern, char equal) {
+    if (!labels) return false;
+
+    struct simple_pattern_match_name_value t = {
+        .pattern = pattern,
+        .equal = equal
+    };
+
+    int ret = dictionary_walkthrough_read(labels, equal?simple_pattern_match_name_and_value_callback:simple_pattern_match_name_only_callback, &t);
+
+    return (ret == -1)?true:false;
 }
 
-struct label *label_list_lookup_keylist(struct label *head, char *key)
-{
-    SIMPLE_PATTERN *pattern = NULL;
+bool rrdlabels_match_simple_pattern(DICTIONARY *labels, const char *simple_pattern_txt) {
+    if (!labels) return false;
 
-    pattern = simple_pattern_create(key, ",|\t\r\n\f\v", SIMPLE_PATTERN_EXACT);
+    SIMPLE_PATTERN *pattern = simple_pattern_create(simple_pattern_txt, " ,|\t\r\n\f\v", SIMPLE_PATTERN_EXACT);
+    char equal = '\0';
 
-    while (head != NULL)
-    {
-        if (simple_pattern_matches(pattern, head->key))
+    const char *s;
+    for(s = simple_pattern_txt; *s ; s++) {
+        if (*s == '=' || *s == ':') {
+            equal = *s;
             break;
-        head = head->next;
+        }
     }
+
+    bool ret = rrdlabels_match_simple_pattern_parsed(labels, pattern, equal);
+
     simple_pattern_free(pattern);
-    return head;
+
+    return ret;
 }
 
-int label_list_contains_keylist(struct label *head, char *keylist)
-{
-    return (label_list_lookup_keylist(head, keylist) != NULL);
+
+// ----------------------------------------------------------------------------
+// Log all labels
+
+static int rrdlabels_log_label_to_buffer_callback(const char *name, void *value, void *data) {
+    BUFFER *wb = (BUFFER *)data;
+    RRDLABEL *lb = (RRDLABEL *)value;
+
+    buffer_sprintf(wb, "Label: %s: \"%s\" (", name, lb->value);
+
+    size_t sources = 0;
+    if(lb->label_source & RRDLABEL_SRC_AUTO)
+        buffer_sprintf(wb, "%sauto", sources++?",":"");
```

Comment:
Hello @ktsaou,

I decided to run `coverity-scan` on your PR due the complexity it has, and I received this report:

```c
*** CID 353213:  Control flow issues  (DEADCODE)
/database/rrdlabels.c: 829 in rrdlabels_log_label_to_buffer_callback()
823         RRDLABEL *lb = (RRDLABEL *)value;
824     
825         buffer_sprintf(wb, "Label: %s: \"%s\" (", name, lb->value);
826     
827         size_t sources = 0;
828         if(lb->label_source & RRDLABEL_SRC_AUTO)
>>>     CID 353213:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach the expression "","" inside this statement: "buffer_sprintf(wb, "%sauto"...".
829             buffer_sprintf(wb, "%sauto", sources++?",":"");
830     
831         if(lb->label_source & RRDLABEL_SRC_CONFIG)
832             buffer_sprintf(wb, "%snetdata.conf", sources++?",":"");
833     
834         if(lb->label_source & RRDLABEL_SRC_K8S)

** CID 353212:  Concurrent data access violations  (MISSING_LOCK)
/mqtt_websockets/src/mqtt_ng.c: 389 in buffer_purge()

````

Reply:
Well, coverity is wrong. Use `netdatacli` to reload the labels to see it working.

Reply:
All right, I am marking as resolved, because the proposed test worked.